2-Draft-Accepted/RFCNNNN-PSCore6-and-Windows-Modules.md

```diff
@@ -0,0 +1,93 @@
+Currently, this property is only documentation and not observed by PowerShell.
+The proposal is to modify PowerShell Core 6.1 to respect this property for in-box Windows PowerShell modules.
+
+### Windows PowerShell 5.1 PSModulePath
+
+Windows PowerShell, by default, searches for modules from these paths in this order:
+
+  1. User profile: $env:UserProfile\Documents\WindowsPowerShell\Modules
+  2. Program files: $env:ProgramFiles\WindowsPowerShell\Modules
+  3. System32: $env:Windir\System32\WindowsPowerShell\v1.0\Modules
+
+### Proposed PowerShell Core PSModulePath Change
+
+PowerShell Core, by default, searches for modules form these paths in this order:
+
+  1. User profile: $env:UserProfile\Documents\PowerShell\Modules
+  2. Program files: $env:ProgramFiles\PowerShell\Modules
+  3. PSHOME: value of $PSHOME
+
+The proposed change is to add *only* the `System32` path from Windows PowerShell 5.1 to `PSModulePath` on PowerShell Core 6
+as the last path in `PSModulePath`.
+
+This means that modules installed using Windows PowerShell with `-Scope CurrentUser` will not be available from PowerShell
+Core 6.
+To use those modules, the user needs to use `Install-Module -Scope CurrentUser` from within PowerShell Core 6.
+Duplicate modules are something that can be addressed in PowerShellGet and outside the scope of this RFC.
+
+Modules in `Program Files` that may be compatible with PowerShell Core 6 is addressed below.
+
+Modules in the PowerShell Core `PSModulePath` are implicitly deemed compatible even if the manifest does not
+declare compatibility.
+A user who explicitly uses `Install-Module` within PowerShell Core 6 expect those modules to be available to PowerShell Core 6.
+
+### CompatiblePSEditions for Windows System32
+
+When we add the `System32` path to PowerShell Core's `PSModulePath`,
+we will use additional logic to validate that the module is compatible with PowerShell Core.
+
+If the module manifest contains `CompatiblePSEditions` with value `Core`, then that module is treated as compatible.
+If the module manifest contains `CompatiblePSEditions` with only the value of `Desktop`, then that module is treated as incompatible.
+If the module manifest does not contain `CompatiblePSEditions`, then that module is treated the same as `Desktop` which is incompatible.
+
+By default, `Get-Module -ListAvailable` will only show modules found in `PSModulePath` that are declared as compatible.
+Similarly, module auto-discovery and auto-loading will only find modules that are declared as compatible.
+
+The `ModuleInfo` table format will add a column for `PSEditions` (shortened from `CompatiblePSEditions` between `Name` and `ExportedCommands`).
+
+### Using Modules not declared as Compatible
+
+`Get-Module -ListAvailable` will support `-SkipCompatibilityCheck` switch to list all modules in `PSModulePath`.
+
+If `Import-Module` is used with a module that is not declared in the manifest as compatible, it will throw a terminating error:
```

Comment:
What is the prescribed behaviour for non-`psd1` modules -- ones that have no manifest?

What should the `PSModuleInfo.PSEdition` fields for those modules say?

Reply:
Since this only applies to modules in `System32`, there shouldn't be any such modules.  If there are, we should add a module manifest.  However, if we missed any then I would expect it to be the same as having a module manifest without `CompatiblePSEditions` property.

Reply:
Maybe I've missed something...

I'm thinking that `Import-Module` and `Get-Module` can be called on any module file, even ones not under `System32`. They can both produce a `PSModuleInfo` object and share code paths with manifest-described modules.

So with the new behaviour described in this RFC, `Import-Module ./my/unpubslished/module.psm1` will have to make some decision on `PSEdition` compatibility settings. It seems tempting to say that no `CompatiblePSEditions` defaults to `@("Desktop")` in general, but that will mean breaking all `psm1`s in Core.

So we need to have a separate behaviour for standalone `psm1`s that the module cmdlets should follow, like just assuming total compatibility.

So the questions are, for a `psm1`:
* Should we always assume compatibility?
* What should `PSModuleInfo.CompatiblePSEditions` contain?
* Does `-SkipCompatibilityCheck` have any function in that scenario?

Reply:
Yes, you found a case not covered explicitly in the RFC which is modules that in PSModulePath.  The intent is that `CompatiblePSEditions` is only enforced for modules under `System32` in the section on line 63.  So in all other cases, it is implicitly deemed compatible.

Reply:
Ah!

I was reading the RFC (in the original version) as:
> Modules in the PowerShell Core PSModulePaths are implicitly deemed compatible even if the manifest does not declare compatibility... If the module manifest contains CompatiblePSEditions with value Core, then that module is treated as compatible. If the module manifest contains CompatiblePSEditions with only the value of Desktop, then that module is treated as incompatible.

But are you saying that it's more specific to the `System32` path:
> All modules **not** found on the `System32` path are implicitly compatibility-whitelisted. The compatibility of modules found on the `System32` path will depend on their respective `CompatiblePSEditions` field.

Is there also intent for gallery module compatibility checking? Or only if they are installed to the `System32` location?